crypto/x509v3/v3_utl.c

```diff
@@ -100,6 +101,45 @@ int X509V3_add_value_bool_nf(const char *name, int asn1_bool,
     return 1;
 }
 
+static char *bignum_to_string(const BIGNUM *bn)
+{
+    char *tmp, *ret;
+    size_t len;
+
+    /*
+     * Display large numbers in hex and small numbers in decimal. Converting to
+     * decimal takes quadratic time and is no more useful than hex for large
+     * numbers.
+     */
+    if (BN_num_bits(bn) < 128) {
+        return BN_bn2dec(bn);
+    }
+
+    tmp = BN_bn2hex(bn);
+    if (tmp == NULL) {
+        return NULL;
+    }
+
+    len = strlen(tmp) + 3;
+    ret = OPENSSL_malloc(len);
+    if (ret == NULL) {
+        X509V3err(X509V3_F_BIGNUM_TO_STRING, ERR_R_MALLOC_FAILURE);
+        OPENSSL_free(tmp);
+        return NULL;
+    }
+
+    /* Prepend "0x", but place it after the "-" if negative. */
+    if (tmp[0] == '-') {
+        OPENSSL_strlcpy(ret, "-0x", len);
+        OPENSSL_strlcat(ret, tmp + 1, len);
+    } else {
+        OPENSSL_strlcpy(ret, "0x", len);
+        OPENSSL_strlcat(ret, tmp, len);
+    }
+    OPENSSL_free(tmp);
+    return ret;
+}
```

Comment:
Unfortunately, this is incredibly memory inefficient.  Considering this seems to be for master only, I wouldn't mind at all if there was a variant if `BN_bn2hex` that was called `BN_bn2hex_buf` with this call pattern:

``` C
    char *buf = NULL;
    size_t siz = BN_bn2hex_buf(bn, buf, 0);     /* ask what buffer size is required, triggered by a NULL buffer */

    if ((buf = OPENSSL_malloc(siz)) == NULL) {
        WHATEVERerr(WHATEVER_F_SOMETHING, ERR_R_MALLOC_FAILURE);
        return NULL;
    }
    BN_bn2hex_buf(bn, buf, siz);

    /* ... */
```

Reply:
If this is meant for 1.1.0 as well, I don't mind if there would appear an *internal* `bn_bn2hex_buf` that could then be made public for 1.1.1

Reply:
It's meant for wherever you want to put it. The issue exists in all OpenSSL versions, including 1.0.2. A certificate with 100k of path length constraint currently takes several seconds to feed into `openssl x509 -text`. And note this scales quadratically, so 5MB will take more than an hour. Whether that's important enough to backport I'll leave up to you. It's just a pretty-printing function, but, then again, folks like crt.sh do like to serve the OpenSSL pretty-print format.

I don't think it makes much sense to be concerned about the single extra malloc + copy here. It is, at the end of the day, a pretty-printing function. Avoiding quadratic scaling is good manners, but it doesn't need to be finely tuned. Besides, getting rid of this copy will not do anything because the entire path is full of this.

Note how each of the callers of this function already start with `ASN1_{ENUMERATED,INTEGER}_to_BN` which produces a copy within the `BIGNUM`. Those callers, in turn, loop over every extension producing these temporary buffers when perhaps they could stream the hex into the `BIO`. There's even an extra temporary copy for `CHARSET_EBCDIC`. And, of course, there's the fundamental crypto/asn1 inefficiency where the `ASN1_INTEGER` itself contains a copy of its contents rather than aliasing into the serialized certificate. Unlike everything else here, that one actually impacts real-world performance in our experience.

Reply:
Yeahok, point taken.  I guess I was overly anal there ;-)

Thanks for taking the time to point out those inefficiencies...  some of them could certainly benefit from handling memory more efficiently (I'll have to look into `ASN1_INTEGER` more closely before I can comment on that one).

Reply:
近30000可以ECB MM NBkSZ用hum徐某与SLi